### PR TITLE
[Ldap][Security] Make LdapAuthenticator an EntryPoint

### LINKED ISSUE #42346: [Ldap] form_login_ldap and new authentication system
**Symfony version(s) affected**: 5.1+

**Description**
With the new authenticator manager enabled, the form_login_ldap authenticator doesn't take care of the login_path option

---------------

**How to reproduce**
- Create new project
- Enable the new authentication system (if not)
- Add the form_login_ldap authenticator:
    ```yaml
        form_login_ldap:
                login_path: security_login
                check_path: security_login
                service: Symfony\Component\Ldap\Ldap
                dn_string: 'uid={username},dc=example,dc=com'
    ```
- Protect one route for authenticated users only:
    ```yaml
        access_control:
        - { path: ^/login$, role: PUBLIC_ACCESS }
        - { path: ^/protected, role: ROLE_USER }
    ```
- Try to access to `/protected`
- You got an Unauthorized exception instead of being redirected to `/login`

If you try to add: 
```yaml
    entry_point: form_login_ldap
```
You'll get a `ServiceNotFoundException`, which suggest to switch to `entry_point: security.authenticator.form_login_ldap.main` which leads to a new issue: 

> Argument 5 passed to Symfony\Component\Security\Http\Firewall\ExceptionListener::__construct() must implement interface Symfony\Component\Security\Http\EntryPoint\AuthenticationEntryPointInterface or be null, instance of Symfony\Component\Ldap\Security\LdapAuthenticator given

---------------------------

**Possible Solution**
Make `Symfony\Component\Ldap\Security\LdapAuthenticator` extends of `AbstractLoginFormAuthenticator`.
This way we got a `getLoginUrl()` methods which resolve our main issue. 

The problem is that, unlike the `Symfony\Component\Security\Http\Authenticator`, we don't have the `HttpUtils` service which prevents us to properly resolve the `login_path` option
I'm not sure on how to inject properly the service to the authenticator so this is why I'm opening an issue instead of doing a MR

### PR DESCRIPTION
**Pull request by dcp-dev**
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #42346
| License       | MIT
| Doc PR        | N/A

I added @chalasr's recommandations given in https://github.com/symfony/symfony/issues/42346#issuecomment-891188282

### CODE REVIEW

**Review by chalasr — CHANGES_REQUESTED** [review]
Thanks for the PR. 1 minor comment.
Sadly we don't have much tests for ldap-based authenticators so I'm ok to merge the patch without additional tests. But if you want to have a look and try to add some, that would be great.

File: src/Symfony/Component/Ldap/Security/LdapAuthenticator.php [hunk 1]
```diff
@@ -75,4 +78,18 @@ public function onAuthenticationFailure(Request $request, AuthenticationExceptio
     {
         return $this->authenticator->onAuthenticationFailure($request, $exception);
     }
+
+    public function start(Request $request, AuthenticationException $authException = null): Response
+    {
+        if (!$this->authenticator instanceof AuthenticationEntryPointInterface) {
+            throw new LogicException(sprintf('Decorated authenticator must implement inerface "%s".', AuthenticationEntryPointInterface::class));
```
Comment (chalasr) [inline src/Symfony/Component/Ldap/Security/LdapAuthenticator.php]:
```suggestion
            throw new LogicException(sprintf('Decorated authenticator "%s" must implement interface "%s".', get_debug_type($this->authenticator), AuthenticationEntryPointInterface::class));
```

**Review by derrabus — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by wouterj** [comment]
Hi! I guess this is a question for @chalasr as he proposed this solution, but I don't think the `LogicException` works/is nice. This means that if the internal authenticator isn't an entry point, a user will get a 500 instead of the default authentication entry point behavior. It's perfectly valid for an authenticator to not be an entry point (e.g. see the `JsonLoginFactory`, which also has a `json_login_ldap` variant). Am I missing something or should we find a different way of approaching this? (maybe a special exception class that we catch when calling the `start()` method?)

**Review by chalasr — CHANGES_REQUESTED** [review]
Just blocking merge to give us some time to address Wouter’s concerns :)

**Comment by chalasr** [comment]
> maybe a special exception class that we catch when calling the start() method?

That's the best approach AFAICT. I just pushed a commit doing that.
Review needed :)

**Comment by chalasr** [comment]
PR ready (I need this for #42582)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @dcp-dev.

**Comment by fabpot** [comment]
@chalasr Merged now :)

**Comment by chalasr** [comment]
Great, thanks!

**Pull request merged**